graph/pull_v2.go

```diff
@@ -232,60 +234,100 @@ func (p *v2Puller) pullV2Tag(tag, taggedName string) (bool, error) {
+			// been skipped due to this layer already existing.
+			continue
 		}
-		if d.layer != nil {
-			// if tmpFile is empty assume download and extracted elsewhere
-			defer os.Remove(d.tmpFile.Name())
-			defer d.tmpFile.Close()
-			d.tmpFile.Seek(0, 0)
-			if d.tmpFile != nil {
-
-				reader := progressreader.New(progressreader.Config{
-					In:        d.tmpFile,
-					Out:       out,
-					Formatter: p.sf,
-					Size:      d.size,
-					NewLines:  false,
-					ID:        stringid.TruncateID(d.img.ID),
-					Action:    "Extracting",
-				})
-
-				err = p.graph.Register(d.img, reader)
-				if err != nil {
-					return false, err
-				}
-
-				if err := p.graph.SetDigest(d.img.ID, d.digest); err != nil {
-					return false, err
-				}
-
-				// FIXME: Pool release here for parallel tag pull (ensures any downloads block until fully extracted)
-			}
-			out.Write(p.sf.FormatProgress(stringid.TruncateID(d.img.ID), "Pull complete", nil))
-			tagUpdated = true
-		} else {
-			out.Write(p.sf.FormatProgress(stringid.TruncateID(d.img.ID), "Already exists", nil))
+
+		// This waits for the download to finish successfully or fail. We need
+		// to wait for all downloads even if previous ones failed so that we
+		// wait for all of the goroutines to die.
+		if err := <-d.err; err != nil {
+			pullErred = true
+			out.Write(p.sf.FormatProgress(stringid.TruncateID(d.img.ID), fmt.Sprintf("Download failed: %v", err), nil))
+			// This download failed. There's nothing we can do. Skip.
+			continue
 		}
+
+		// The download finished successfully. A tmpFile field is *always* set
+		// on successful downloads.
+		defer os.Remove(d.tmpFile.Name())
+		defer d.tmpFile.Close()
+
+		if pullErred {
+			// There was an error downloading or installing one or more
+			// ancestor filesystem layers. We cannot continue extracting this
+			// one because it depends on the parent layer being installed first.
+			out.Write(p.sf.FormatProgress(stringid.TruncateID(d.img.ID), "Unable to install filesystem layer: failed to install ancestor layer", nil))
+			continue
+		}
+
+		// Seek to the beginning of the downloaded layer archive to begin
+		// extracting it.
+		d.tmpFile.Seek(0, 0)
+		reader := progressreader.New(progressreader.Config{
+			In:        d.tmpFile,
+			Out:       out,
+			Formatter: p.sf,
+			Size:      d.size,
+			NewLines:  false,
+			ID:        stringid.TruncateID(d.img.ID),
+			Action:    "Extracting",
+		})
+
+		err = p.graph.Register(d.img, reader)
+		if err != nil {
+			pullErred = true
+			out.Write(p.sf.FormatProgress(stringid.TruncateID(d.img.ID), fmt.Sprintf("Unable to install filesystem layer: %v", err), nil))
+			continue
+		}
+
+		if err := p.graph.SetDigest(d.img.ID, d.digest); err != nil {
+			pullErred = true
+			out.Write(p.sf.FormatProgress(stringid.TruncateID(d.img.ID), fmt.Sprintf("Unable to save filesystem layer digest: %v", err), nil))
+			continue
+		}
+
+		// FIXME: Pool release here for parallel tag pull (ensures any
```

Comment:
There is a possibility that a pool lock was taken above that needs to be released in this section. I'm not sure what we mean by "parallel tag pull".